src/secp256k1.c

```diff
@@ -681,10 +730,144 @@ int secp256k1_ec_pubkey_combine(const secp256k1_context* ctx, secp256k1_pubkey *
     return 1;
 }
 
+/* Converts the point encoded by a secp256k1_pubkey into its absolute value,
+ * i.e. keeps it as is if it is positive and otherwise negates it. Sign is set
+ * to 1 if the input point was negative and set to 0 otherwise. */
+static void secp256k1_ec_pubkey_absolute(const secp256k1_context* ctx, secp256k1_pubkey *pubkey, int *sign) {
+   secp256k1_ge ge;
+    secp256k1_pubkey_load(ctx, &ge, pubkey);
+    if (sign != NULL) {
+        *sign = 0;
+    }
+    if (!secp256k1_fe_is_quad_var(&ge.y)) {
+        secp256k1_ge_neg(&ge, &ge);
+        if (sign != NULL) {
+            *sign = 1;
+        }
+    }
+    secp256k1_pubkey_save(pubkey, &ge);
+}
+
+int secp256k1_xonly_pubkey_create(const secp256k1_context* ctx, secp256k1_xonly_pubkey *pubkey, const unsigned char *seckey) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(seckey != NULL);
+
+    if (!secp256k1_ec_pubkey_create(ctx, (secp256k1_pubkey *) pubkey, seckey)) {
+        return 0;
+    }
+    secp256k1_ec_pubkey_absolute(ctx, (secp256k1_pubkey *) pubkey, NULL);
+    return 1;
+}
+
+int secp256k1_xonly_pubkey_parse(const secp256k1_context* ctx, secp256k1_xonly_pubkey* pubkey, const unsigned char *input32) {
+    /* TODO parse directly from 32 byte buffer */
+    unsigned char buf[33];
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(input32 != NULL);
+
+    buf[0] = SECP256K1_TAG_PUBKEY_EVEN;
+    memcpy(&buf[1], input32, 32);
+    if (!secp256k1_ec_pubkey_parse(ctx, (secp256k1_pubkey *) pubkey, buf, sizeof(buf))) {
+        return 0;
```

Comment:
As it stands, if `secp256k1_xonly_pubkey_parse` fails, it will leave the `pubkey` value in a invalid state.  If the user of the this API were to ignore the return value of `secp256k1_xonly_pubkey_parse` and proceeded to call `secp256k1_schnorrsig_verify`, the verification could, perhaps, succeed.

However if we were to add a line such as `memset(pubkey, 0, sizeof(secp256k1_xonly_pubkey);` here, then `pubkey` would be initialized to a value that denotes a point with a zero x-coordinate.  The current implementation of `secp256k1_schnorrsig_verify` calls `secp256k1_pubkey_load` which explicitly fails when passed a pubkey point with a zero x-coordinate, which in turn would cause `secp256k1_schnorrsig_verify` to unconditionally fail, even if the result of `secp256k1_xonly_pubkey_parse` were ignored.  (Not only does `secp256k1_pubkey_load` unconditionally fail, but it also calls `ctx->illegal_callback`).

Reply:
This is a valid concern. But `secp256k1_ec_pubkey_parse` does that already:
https://github.com/bitcoin-core/secp256k1/blob/e541a90ef6461007d9c6a74b9f9a7fb8aa34aaa8/src/secp256k1.c#L256

Reply:
I added a commit with an explicit test that the pubkey is zeroized.

Reply:
@real-or-random Thanks.  I missed the indirection in `secp256k1_ec_pubkey_parse` between `Q` and `pubkey`.